apps/docs/pages/guides/ai/vector-indexes/hnsw-indexes.mdx

```diff
@@ -0,0 +1,101 @@
+| -------- | ---------------------- | ------------------------------------------------------------------------------------ |
+| `<->`    | Euclidean distance     | `vector_l2_ops`                                                                      |
+| `<#>`    | negative inner product | `vector_ip_ops`                                                                      |
+| `<=>`    | cosine distance        | `vector_cosine_ops`                                                                  |
+
+Use the following SQL commands to create an HNSW index for the operator(s) used in your queries.
+
+### Euclidean L2 distance (`vector_l2_ops`)
+
+```sql
+create index on items using hnsw (column_name vector_l2_ops);
+```
+
+### Inner product (`vector_ip_ops`)
+
+```sql
+create index on items using hnsw (column_name vector_ip_ops);
+```
+
+### Cosine distance (`vector_cosine_ops`)
+
+```sql
+create index on items using hnsw (column_name vector_cosine_ops);
+```
+
+Currently vectors with up to 2,000 dimensions can be indexed.
+
+If you are using the `vecs` Python library, follow the instructions in [Managing collections](/docs/guides/ai/managing-collections#create-an-index) to create indexes.
+
+## How does HNSW work?
+
+HNSW uses proximity graphs (graphs connecting nodes based on distance between them) to approximate nearest-neighbor search. To understand HNSW, we can break it down into 2 parts:
+
+- **Hierarchical (H):** The algorithm operates over multiple layers
+- **Navigable Small World (NSW):** Each vector is a node within a graph and is connected to several other nodes
+
+### Hierarchical
+
+The hierarchical aspect of HNSW builds off of the idea of skip lists.
+
+Skip lists are multi-layer linked lists. The bottom layer is a regular linked list connecting an ordered sequence of elements. Each new layer above removes some elements from the underlying layer (based on a fixed probability), producing a sparser subsequence that “skips” over elements.
+
+<div>
+  <img
+    alt="visual of an example skip list (light)"
+    className="dark:hidden"
+    src="/docs/img/ai/vector-indexes/hnsw-indexes/skip-list--light.png"
+  />
+  <img
+    alt="visual of an example skip list (dark)"
+    className="hidden dark:block"
+    src="/docs/img/ai/vector-indexes/hnsw-indexes/skip-list--dark.png"
+  />
+</div>
+
+When searching for an element, the algorithm begins at the top layer and traverses its linked list horizontally. If the target element is found, the algorithm stops and returns it. Otherwise if the next element in the list is greater than the target (or NIL), the algorithm drops down to the next layer below. Since each layer below is less sparse than the layer above (with the bottom layer connecting all elements), the target will eventually be found. Skip lists offer O(log n) average complexity for both search and insertion/deletion.
```

Comment:
```suggestion
When searching for an element, the algorithm begins at the top layer and traverses its linked list horizontally. If the target element is found, the algorithm stops and returns it. Otherwise if the next element in the list is greater than the target (or NULL), the algorithm drops down to the next layer below. Since each layer below is less sparse than the layer above (with the bottom layer connecting all elements), the target will eventually be found. Skip lists offer O(log n) average complexity for both search and insertion/deletion.
```